web/auth/interceptors/access_control.go

```diff
@@ -0,0 +1,134 @@
+package interceptors
+
+import (
+	"context"
+	"strings"
+
+	pb "github.com/autograde/quickfeed/ag"
+	"github.com/autograde/quickfeed/web/auth"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// "user" role implies that user attempts to access information about himself.
+// "group" role implies that the user is a course student + a member of the given group.
+// "student" role implies that the user is enrolled in the course with any role.
+// "teacher": user enrolled in the course with teacher status.
+// "owner" is the creator of the manual review.
+var accessRoles = map[string][]string{
+	"UpdateUser":              {"user", "admin"},
+	"GetEnrollmentsByUser":    {"user", "admin"},
+	"GetEnrollmentsByCourse":  {"student"},
+	"UpdateReview":            {"owner", "teacher"}, // TODO(vera): a new role for review owner or just a server-side check?
+	"GetGroupByUserAndCourse": {"group", "teacher"},
+	"CreateGroup":             {"group", "teacher"},
+	"GetGroup":                {"group", "teacher"}, // TODO(vera): "group" needs a db call or a field in claims
+	"UpdateGroup":             {"teacher"},
+	"DeleteGroup":             {"teacher"},
+	"GetSubmissions":          {"teacher"},
+	"IsEmptyRepo":             {"teacher"},
+	"GetGroupsByCourse":       {"teacher"},
+	"UpdateCourse":            {"teacher"},
+	"UpdateEnrollments":       {"teacher"},
+	"UpdateSubmission":        {"teacher"},
+	"RebuildSubmissions":      {"teacher"},
+	"CreateBenchmark":         {"teacher"},
+	"UpdateBenchmark":         {"teacher"},
+	"DeleteBenchmark":         {"teacher"},
+	"CreateCriterion":         {"teacher"},
+	"UpdateCriterion":         {"teacher"},
+	"DeleteCriterion":         {"teacher"},
+	"CreateReview":            {"teacher"},
+	"UpdateSubmissions":       {"teacher"},
+	"GetReviewers":            {"teacher"},
+	"UpdateAssignments":       {"teacher"},
+	"GetUserByCourse":         {"teacher", "admin"},
+	"GetSubmissionsByCourse":  {"teacher", "admin"},
+	"GetOrganization":         {"admin"},
+	// TODO(vera): GetOrganization and other admin methods not needed in case of admin service
+	// otherwise add all admin methods here
+}
+
+func AccessControl(logger *zap.Logger, tokens *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		// get JWT from context
+		meta, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			logger.Sugar().Errorf("Access control failed: missing metadata")
+			return nil, status.Errorf(codes.Unauthenticated, "access denied")
+		}
+		logger.Sugar().Debugf("Request %s has metadata: %+v", info.FullMethod, meta) // tmp
+		// get claims from jwt
+		for _, c := range meta.Get("cookie") {
```

Comment:
There is a helper function `auth.extractToken` that does more or less the same thing. Didn't check if we can or should export that function, but we should at least avoid many nested loops and ifs -- it makes the code difficult to read.